pkg/controller/azvolume.go

```diff
@@ -81,260 +82,303 @@ func (r *ReconcileAzVolume) Reconcile(ctx context.Context, request reconcile.Req
-		return nil
+	return nil
+}
+
+func (r *ReconcileAzVolume) triggerRelease(ctx context.Context, volumeName string, azVolume *v1alpha1.AzVolume) error {
+	if _, err := cleanUpAzVolumeAttachmentByVolume(ctx, r, volumeName, replicaOnly); err != nil {
+		return err
+	}
+	updateFunc := func(azv *v1alpha1.AzVolume) {
+		_ = r.updateStatusPhase(ctx, azv, v1alpha1.VolumeAvailable)
 	}
 
-	// Update status of the object
-	if _, err = r.updateStatus(ctx, azVolume.Name, azVolume, v1alpha1.VolumeReleased, true, nil); err != nil {
+	if err := r.updateWithRetry(ctx, volumeName, updateFunc); err != nil {
 		return err
 	}
 
-	klog.Infof("successfully deleted volume (%s) and its attachments and update status of AzVolume (%s)", azVolume.Spec.UnderlyingVolume, azVolume.Name)
 	return nil
 }
 
-func (r *ReconcileAzVolume) updateState(ctx context.Context, volumeName string, azVolume *v1alpha1.AzVolume, state v1alpha1.AzVolumeState) (*v1alpha1.AzVolume, error) {
-	var err error
+func (r *ReconcileAzVolume) initializeMeta(ctx context.Context, azVolume *v1alpha1.AzVolume) *v1alpha1.AzVolume {
 	if azVolume == nil {
-		azVolume, err = azureutils.GetAzVolume(ctx, r.client, r.azVolumeClient, volumeName, r.namespace, true)
-		if err != nil {
-			klog.Errorf("failed to get AzVolume (%s): %v", volumeName, err)
-			return nil, err
-		}
+		return nil
+	}
+	if finalizerExists(azVolume.Finalizers, azureutils.AzVolumeFinalizer) {
+		return azVolume
+	}
+	// add finalizer
+	if azVolume.ObjectMeta.Finalizers == nil {
+		azVolume.ObjectMeta.Finalizers = []string{}
+	}
+	azVolume.ObjectMeta.Finalizers = append(azVolume.ObjectMeta.Finalizers, azureutils.AzVolumeFinalizer)
+
+	return azVolume
+}
 
+func (r *ReconcileAzVolume) deleteFinalizer(ctx context.Context, azVolume *v1alpha1.AzVolume, finalizersToDelete map[string]bool) *v1alpha1.AzVolume {
+	if azVolume == nil {
+		return nil
 	}
 
-	updated := azVolume.DeepCopy()
-	if updated.Status.State == state {
-		return azVolume, nil
+	if azVolume.ObjectMeta.Finalizers == nil {
+		return azVolume
 	}
-	updated.Status.State = state
 
-	if err := r.client.Update(ctx, updated, &client.UpdateOptions{}); err != nil {
-		klog.Errorf("failed to update AzVolume (%s): %v", volumeName, err)
-		return nil, err
+	finalizers := []string{}
+	for _, finalizer := range azVolume.ObjectMeta.Finalizers {
+		if exists := finalizersToDelete[finalizer]; exists {
+			continue
+		}
+		finalizers = append(finalizers, finalizer)
 	}
-	return updated, nil
+	azVolume.ObjectMeta.Finalizers = finalizers
+	return azVolume
 }
 
-func (r *ReconcileAzVolume) updateStatus(ctx context.Context, volumeName string, azVolume *v1alpha1.AzVolume, phase v1alpha1.AzVolumePhase, isDeleted bool, status *v1alpha1.AzVolumeStatusParams) (*v1alpha1.AzVolume, error) {
-	var err error
+func (r *ReconcileAzVolume) updateState(ctx context.Context, azVolume *v1alpha1.AzVolume, state v1alpha1.AzVolumeState) *v1alpha1.AzVolume {
```

Comment:
This function should verify that the requested state is valid for the current state.